### ISSUE TITLE
`pathFilter` (or some new API) should support implementations of `exports`

### ISSUE THREAD

**Issue by SimenB** [browserify/resolve#253]
In lieu of this module supporting `exports` out of the box, it would be awesome if consumers could use its existing capabilities (`packageFilter`, `pathFilter` etc.) to implement support for `exports` on the consumer side.

I have experimented a bit with this, and from what I can tell there are 2 issues:

1. `resolve` adds `index` if a directory is specified (https://github.com/browserify/resolve/blob/f1b51848ecb7f56f77bfb823511d032489a13eab/lib/sync.js#L192). This can potentially cause issues in the case `index` is _not_ specified by the user as changing it to `.` (which is what a directory means in `exports`) is not necessarily a safe operation. Not sure how to mitigate this in a way that's backwards compatible.
2. I tried to use `pathFilter` which works great for `require('some-module/thing')`, but it doesn't work for `require('./some-thing')` from within a module. Relative imports within a module isn't restricted by `exports`, but `pathFilter` cannot know if the resolution request is for a relative file or not. A solution here is maybe to include a flag saying if it's relative within the package or not?

**Comment by SimenB** [comment]
1. True about it becoming obsolete once native supports lands, but it could be removed in a v2 or something? And the logic itself would be kept, just no way of manually switching it on or off

2. `pathFilter` - no leading dot, so it has no way of knowing if the request was `require('module/file')` or `require('./file')` from within `module/index.js` for instance. The `relativePath` arg passed in this case is just `file`. It could be `./file` for module relative imports, and `file` for "external"? That's breaking tho, so for backwards compat I guess we'd need a new arg or something saying if it's one or the other. Or possibly, since `resolve` only works on a single `from` and `to` at the time, I could look at the `basedir` that I pass, and based on that differ in the logic? That might work 🙂

**Comment by ljharb** [comment]
Seems fine - part of the reason `resolve`'s implementation has taken so long is that without gobs and gobs of test cases, it's hard for me to be confident I've gotten it right (and almost every time i've added a fixture project, i find a way i got it wrong) - so you may want to look into adding lots of tests :-)

**Comment by SimenB** [comment]
Right, it breaks a test, but I think that test is faulty (it imports a "local" module with a package.json with `exports` via a relative path, but that is not in `node_modules`). I think that's just wrong behavior, so it failing now is actually a bug fix and my old test was just bad 😅 

Might just roll this out behind a config flag and hope people test it and report instances where it behaves wrong (after adding more tests)

**Comment by SimenB** [comment]
`pathFilter` is working out well 👍

Would be nice to have some way of disabling `main` and the `index.js` in root stuff if there is `exports`, but I guess doing that in isolation doesn't make much sense. And ignoring it manually isn't too bad

**Comment by ljharb** [comment]
Once we have true "exports" support, it'll take precedence over the "main" anyways, so you won't need to ignore that.

**Comment by SimenB** [comment]
Yeah, figured it'd make more sense to go straight there instead of some intermediary step 🙂

**Comment by SimenB** [comment]
Ok, closing again 🙈 My solution is to pop of anything trailing after `pkg/` or `@scope/pkg/`, add `./package.json` and resolve that using `resolve`, and if `package.json` has `exports` resolve those and return the full path. No `packageFilter` or `pathFilter` at all. In theory a tiny bit faster as well as we head straight for the `package.json` file and potentially shortcut instead of traversing and checking for file existence.

(code is in above PR if this explanation made no sense, which it probably doesn't)
